### PR TITLE
Add support for triple-channel INA3221 i2c voltage-and-current sensor

### PR DESCRIPTION
**Pull request by peterbarker**
This adds support for a new i2c-based voltage/current sensor.

This driver is not currently suitable for merging.  It might be that this is more of a component than an e.g. battery monitor in-and-of itself.  It might form a base class for a product based around it, perhaps (as several of the smbus-based devices are based around a "generic" driver).  I'm waiting on more design details to see where it goes from here.

This has been tested on a `CJMCU-3221` breakout board.

Of interest here is the i2c simulator for the device - we're still fleshing out how simulated i2c devices work, but the "writable registers" mask is interesting.

This work is being sponsored by Wurzbach Electronics.

### CODE REVIEW

**Comment by apache405** [comment]
YAY!  

Thanks for taking on the task of getting this driver built.  I'm gonna have a lot of fun putting this into new stuff.

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
+#pragma once
+
+/*
+ *
+
+ * DataSheet: https://www.ti.com/lit/ds/symlink/ina3221.pdf?ts=1597369254046
+ */
+
+// The INA3221takestwo measurementsfor eachchannel:one for shuntvoltageand one for bus voltage.Eachmeasurementcan be independentlyor sequentiallymeasured,basedon the modesetting(bits2-0 in theConfigurationregister).Whenthe INA3221is in normaloperatingmode(thatis, the MODEbits of theConfigurationregisterare set to 111),the devicecontinuouslyconvertsa shunt-voltagereadingfollowedby abus-voltagereading.Thisprocedureconvertsone channel,and thencontinuesto the shuntvoltagereadingofthe nextenabledchannel,followedby the bus-voltagereadingfor that channel,and so on, untilall enabledchannelshavebeenmeasured.The programmedConfigurationregistermodesettingappliesto all channels.Any channelsthat are not enabledare bypassedin the measurementsequence,regardlessof modesetting.
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h:9]: Line breaks, please? I'm not one to normally complain, but not only is this missing line breaks, it's missing a ton of spaces, and is just very hard to read.

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+void AP_BattMonitor_INA3221::init()
+{
+    // check to see which address we should start probing at:
+    uint8_t probe_address = 64;
+    for (uint8_t i = 0; i < AP::battery()._num_instances; i++) {
+        if (AP::battery().get_type(i) != _type) {
+            // different type...
+            continue;
+        }
+        const AP_BattMonitor_INA3221 *driver = ((AP_BattMonitor_INA3221*)AP::battery().drivers[i]);
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp:165]: This isn't really a stable sort for these. An address register would be far better (then deconflict later ones if they are a reuse).

The specific problem I'm thinking of here is lets say battery monitor 1 and 2 are both INA3221 devices, and you remove the first one from the system in hardware. The second one's hardware will now be detected as the first monitor and use it's failsafes which are all incorrect. (This is also just bad if the device isn't detected for some reason)

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        if (!init_at_address(probe_address)) {
+            continue;
+        }
+        break;
+    }
+}
+
+void AP_BattMonitor_INA3221::read()
+{
+    static uint32_t last_print;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp:186]: Presumably this is all debug and will get removed?

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    if (!read_register(reg_shunt, shunt_voltage)) {
+        return;
+    }
+    uint16_t bus_voltage;
+    if (!read_register(reg_bus, bus_voltage)) {
+        return;
+    }
+
+    //transfer readings to front end:
+    WITH_SEMAPHORE(_sem);
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp:150]: So you aren't wrong that there should be a semaphore around writing to the front end, as the SMBus drivers have this bug, but as far as I can tell you have gained no protection here, as you only grab this in your init, and in the timer. It doesn't actually get locked by the frontend ever when reading state out of the battery, so I think as it stands this is wasting processing time/memory. If you want to use this, you need to accumulate the results in your own internal copy here, and then take this semaphore in the `read()` call which currently has your debug print.

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    default:
+        return;
+    }
+    uint16_t shunt_voltage;
+    if (!read_register(reg_shunt, shunt_voltage)) {
+        return;
+    }
+    uint16_t bus_voltage;
+    if (!read_register(reg_bus, bus_voltage)) {
+        return;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp:146]: Do you really want to lose the new/best shunt voltage data you have available if you didn't get the bus voltage? Seems like accumulating some of the data (even if the time stamp isn't moved on) is probably safer for the system.

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+    }
+
+    //transfer readings to front end:
+    WITH_SEMAPHORE(_sem);
+    _state.voltage = bus_voltage;
+    _state.current_amps = shunt_voltage * 0.56f;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp:152]: Why not impelent integration of the current here? (you can steal the math from the analog back end)

Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp:152]: Also where does the 0.56 come from. Based on a quick look at the data sheet I'd have guessed that this is actually specific to the board that you had the INA3221 on, not that all shunts would have this value?

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    case AP_BattMonitor::Type::INA3221_CH2:
+        reg_shunt = 3;
+        reg_bus = 4;
+        break;
+    case AP_BattMonitor::Type::INA3221_CH3:
+        reg_shunt = 5;
+        reg_bus = 6;
+        break;
+    default:
+        return;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp:138]: Internal error, flow of control?

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+private:
+
+    bool init_at_address(uint8_t addr);
+
+    HAL_Semaphore _sem;
+    AP_HAL::OwnPtr<AP_HAL::I2CDevice> _dev;
+
+    const AP_BattMonitor::Type _type;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h:45]: Why do you need the extra copy of this? It's already available for you inside `_params._type` which saves you the memory?

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    uint8_t address;
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h:46]: Unused?

File: libraries/SITL/SIM_INA3221.cpp [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        }
+        if (update_shunt) {
+            float fakecurrent = 7.6;
+            registers.byname.Channel_1_Shunt_Voltage = fakecurrent/0.56; // FIXME
+        }
+    }
+
+    if (registers.byname.configuration.bits.ch2_enable) {
+        if (update_shunt) {
+            registers.byname.Channel_2_Shunt_Voltage = AP::sitl()->state.battery_current/26 * 32768; // FIXME
```
Comment (WickedShell) [inline libraries/SITL/SIM_INA3221.cpp:119]: FIXME's that seem obvious to the author, are not obvious to reviewers immediately, or at future dates, can we get a real comment on what should be fixed?

File: libraries/SITL/SIM_INA3221.cpp [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        }
+        if (update_bus) {
+            registers.byname.Channel_2_Bus_Voltage = AP::sitl()->state.battery_voltage/26 * 32768; // FIXME
```
Comment (WickedShell) [inline libraries/SITL/SIM_INA3221.cpp:122]: Same as above fixme.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 12]
```diff
@@ -185,6 +186,30 @@ AP_BattMonitor::init()
             case Type::Generator:
                 drivers[instance] = new AP_BattMonitor_Generator(*this, state[instance], _params[instance]);
                 break;
+            case Type::INA3221_CH1:
+//                _params[instance]._i2c_bus.set_default(AP_BATTMONITOR_SMBUS_BUS_INTERNAL),
+                drivers[instance] = new AP_BattMonitor_INA3221(
+                    *this,
+                    state[instance],
+                    _params[instance],
+                    Type::INA3221_CH1);
```
Comment (WickedShell) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:195]: As noted before the type really doesn't look to be needed here, and then the arguments can all fit on one line the same as the others around it.

**Review by WickedShell — DISMISSED** [review]

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp (hunk 6 above)
Comment (tpwrules) [reply libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp]: Done

Comment (tpwrules) [reply libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp]: Fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp (hunk 7 above)
Comment (tpwrules) [reply libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp:228]: outdates

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h (hunk 8 above)
Comment (tpwrules) [reply libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h]: outdated

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h (hunk 9 above)
Comment (tpwrules) [reply libraries/AP_BattMonitor/AP_BattMonitor_INA3221.h]: outdated

File: libraries/SITL/SIM_INA3221.cpp (hunk 10 above)
Comment (tpwrules) [reply libraries/SITL/SIM_INA3221.cpp:128]: conversion values need to be fixed, they are known good in the main driver

File: libraries/SITL/SIM_INA3221.cpp (hunk 11 above)
Comment (tpwrules) [reply libraries/SITL/SIM_INA3221.cpp:131]: conversion values need to be fixed, they are known good in the main driver

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp (hunk 12 above)
Comment (tpwrules) [reply libraries/AP_BattMonitor/AP_BattMonitor.cpp]: fixed

**Comment by tpwrules** [comment]
Tested that voltage, current, and accumulation now work properly with real hardware on stampfly.

Please feel free to rebase/squash as desired, though I would like credit for both of us to be preserved.

The sim needs fixup to use the scaling values from the driver, and possibly modernization?

File: libraries/SITL/SIM_INA3221.cpp (hunk 10 above)
Comment (tpwrules) [reply libraries/SITL/SIM_INA3221.cpp:128]: Fixed

File: libraries/SITL/SIM_INA3221.cpp (hunk 11 above)
Comment (tpwrules) [reply libraries/SITL/SIM_INA3221.cpp:131]: Fixed

File: libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp (hunk 4 above)
Comment (tpwrules) [reply libraries/AP_BattMonitor/AP_BattMonitor_INA3221.cpp]: Done

**Comment by tpwrules** [comment]
Okay, this should be ready to go! Tests now do something useful too.

File: libraries/AP_BattMonitor/AP_BattMonitor.cpp [hunk 13]
```diff
@@ -604,6 +605,16 @@ AP_BattMonitor::init()
...
                 break;
 #endif // AP_BATTERY_SCRIPTING_ENABLED
+#if AP_BATTERY_INA3221_ENABLED
+            case Type::INA3221:
+//                _params[instance]._i2c_bus.set_default(AP_BATTMONITOR_SMBUS_BUS_INTERNAL),
+                drivers[instance] = NEW_NOTHROW AP_BattMonitor_INA3221(
+                    *this,
+                    state[instance],
+                    _params[instance]
+                );
```
Comment (srmainwaring) [inline libraries/AP_BattMonitor/AP_BattMonitor.cpp:615]: Format to single line to be consistent with other entries.

File: Tools/autotest/ardusub.py [hunk 14]
```diff
@@ -947,6 +947,50 @@ def FuseMag(self):
...
+            "BATT4_CHANNEL": 3,
+        })
+        self.reboot_sitl()
+
+        seen_1 = False
+        seen_3 = False
+        tstart = self.get_sim_time()
+        while not (seen_1 and seen_3):
+            m = self.assert_receive_message('BATTERY_STATUS')
+            if self.get_sim_time() - tstart > 10:
```
Comment (srmainwaring) [inline Tools/autotest/ardusub.py:978]: 10s results in a timeout when running at standard speed on a mac M1. With `--speedup 2` the test completes. Not sure what the policy is for autotest (do we assume that in CI they'll be run with a speedup?). Increasing to 15 passes.

File: Tools/autotest/ardusub.py [hunk 15]
```diff
@@ -947,6 +947,50 @@ def FuseMag(self):
...
+                raise NotAchievedException("INA3221 status timeout")
+            if m.id == 1:
+                self.assert_message_field_values(m, {
+                    "current_battery": int(3.14159 * 100),
```
Comment (srmainwaring) [inline Tools/autotest/ardusub.py:982]: Additional documentation of the choice of test value for each channel (hardcoded vs simulated) here and in SIM_INA3221.h/.cpp would help future maintainers.

**Review by srmainwaring — CHANGES_REQUESTED** [review]
Looks good. I have flown the stampfly branch containing these changes with ELRS telem and the reported voltage and current seem reasonable.

I have not gone through the driver in detail (comparison with datasheet etc.). If you'd like that carried out as part of the review pls ltm.

Have a few suggestions:
- increase test timeout so it completes without needing to enable speedup 
- add further comments on the hardcoded test values in test and simulated sensor

File: Tools/autotest/ardusub.py (hunk 14 above)
Comment (tpwrules) [reply Tools/autotest/ardusub.py:978]: This is strange to hear, I chose 10 seconds as a "couldn't possibly fail" value. There might be an inherent test timeout too which would remove the need to check, but I'm not 100% sure? Also `get_sim_time` should be scaled with speedup.

How are you running the test? Using `Tools/autotest/autotest.py build.Sub test.Sub.INA3221 --no-clean` and a print statement the total elapsed "time" is well under 1 second on my x86 machine and passes every time. I tested using `--speedup=1`, `--speedup=2`, and whatever the autotest script default is.

Comment (srmainwaring) [reply Tools/autotest/ardusub.py:978]: Disregard - I was using a `--debug` build and that explains the timeout.

File: Tools/autotest/ardusub.py (hunk 15 above)
Comment (tpwrules) [reply Tools/autotest/ardusub.py:982]: Fixed

**Comment by tridge** [comment]
@tpwrules @peterbarker this PR looks good to me, just needs some commits squashing

**Review by srmainwaring — APPROVED** [review]

**Review by tridge — APPROVED** [review]

**Pull request merged**